### PR TITLE
SE-0054: Rework diagnostics for IUOs and revise Swift 3 /4 semantics.

### PR DESCRIPTION
**Pull request by rudkx**
For Swift 3 / 4:

Deprecate the spelling "ImplicitlyUnwrappedOptional", emitting a warning
and suggesting "!" in places where they are allowed according to
SE-0054.

In places where SE-0054 disallowed IUOs but we continued to accept them
in previous compilers, emit a warning suggesting "Optional" or "?"  as
an alternative depending on context and treat the IUO as an Optional,
noting this in the diagnostic.

For Swift 5:

Treat "ImplicitlyUnwrappedOptional" as an error, suggesting
"!" in places where they are allowed by SE-0054.

In places where SE-0054 disallowed IUOs, emit an error suggestion
"Optional" or "?" as an alternative depending on context.

### CODE REVIEW

**Comment by rudkx** [comment]
@swift-ci Please smoke test

**Comment by rudkx** [comment]
@swift-ci Please test source compatibility

**Comment by rudkx** [comment]
@swift-ci Please smoke test

**Comment by rudkx** [comment]
@swift-ci Please test source compatibility

**Comment by rudkx** [comment]
@swift-ci Please smoke test

**Comment by rudkx** [comment]
@swift-ci Please test source compatibility

File: include/swift/AST/DiagnosticsSema.def [hunk 1]
```diff
@@ -3104,20 +3104,31 @@ ERROR(tuple_ellipsis,none,
       "cannot create a variadic tuple", ())
 
 WARNING(implicitly_unwrapped_optional_spelling_deprecated,none,
-        "the spelling 'ImplicitlyUnwrappedOptional' is deprecated", ())
+        "the spelling 'ImplicitlyUnwrappedOptional' in deprecated; use an explicit type followed by '!'", ())
```
Comment (xwu) [inline include/swift/AST/DiagnosticsSema.def]: I think you still mean "_is_ deprecated"?

Comment (rudkx) [reply include/swift/AST/DiagnosticsSema.def]: Indeed, thanks.

File: include/swift/AST/DiagnosticsSema.def [hunk 2]
```diff
@@ -3104,20 +3104,31 @@ ERROR(tuple_ellipsis,none,
...
         "the spelling 'ImplicitlyUnwrappedOptional' is deprecated; use '!' after the type name", ())
 
+WARNING(implicitly_unwrapped_optional_spelling_decay_to_optional,none,
+        "'ImplicitlyUnwrappedOptional' is not allowed here; interpreting this as 'Optional' instead", ())
+
+WARNING(implicitly_unwrapped_optional_in_illegal_position_decay_to_optional,none,
+        "'!' is not allowed here; interpreting this as '?' instead", ())
+
 ERROR(implicitly_unwrapped_optional_spelling_error,none,
         "the spelling 'ImplicitlyUnwrappedOptional' in unsupported; use an explicit type followed by '!'", ())
```
Comment (rudkx) [inline include/swift/AST/DiagnosticsSema.def]: @xwu Fixed this as well.

**Comment by rudkx** [comment]
@swift-ci Please smoke test

**Comment by rudkx** [comment]
@swift-ci Please test source compatibility

**Comment by rudkx** [comment]
@swift-ci Please test source compatibility

**Comment by rudkx** [comment]
@nathawes Any ideas on why I am seeing this behavior in test/Migrator/wrap_optional.swift?

Is this indicating that the migrator is not actually doing anything on this code with the changes that I have here?

**Comment by nathawes** [comment]
Not sure about this but my guess is that the change mentioned in the warnings from the logs:
```
/Users/buildnode/jenkins/workspace/swift-PR-osx-smoke-test/branch-master/swift/test/Migrator/Inputs/Cities.swift:19:39: warning: '!' is not allowed here; interpreting this as '?' instead
  func blibli(x: (String?, String) -> String!)
                                      ^     ~
                                            ?
```
Might also change the USR of that symbol so it no longer matches the USR the migration is matching against from `test/Migrator/Inputs/API.json`:
```
  {
    "DiffItemKind": "CommonDiffItem",
    "NodeKind": "Function",
    "NodeAnnotation": "WrapOptional",
    "ChildIndex": "1",
    "LeftUsr": "s:6Cities05ExtraA0P6blibliySQySSGSSSg_SStc1x_tF",
    "LeftComment": "",
    "RightUsr": "",
    "RightComment": "",
    "ModuleName": "Cities"
  },
  {
    "DiffItemKind": "CommonDiffItem",
    "NodeKind": "Function",
    "NodeAnnotation": "WrapOptional",
    "ChildIndex": "1",
    "LeftUsr": "s:6Cities05ExtraA0P6blibliySQySSGSSSg_SStc1x_tF",
    "LeftComment": "",
    "RightUsr": "",
    "RightComment": "",
    "ModuleName": "Cities"
  },
  {
    "DiffItemKind": "CommonDiffItem",
    "NodeKind": "Function",
    "NodeAnnotation": "WrapOptional",
    "ChildIndex": "1",
    "LeftUsr": "s:6Cities05ExtraA0P6blibliySQySSGSSSg_SStc1x_tF",
    "LeftComment": "",
    "RightUsr": "",
    "RightComment": "",
    "ModuleName": "Cities"
  },
  {
    "DiffItemKind": "CommonDiffItem",
    "NodeKind": "Function",
    "NodeAnnotation": "ImplicitOptionalToOptional",
    "ChildIndex": "1:0",
    "LeftUsr": "s:6Cities05ExtraA0P6blibliySQySSGSSSg_SStc1x_tF",
    "LeftComment": "",
    "RightUsr": "",
    "RightComment": "",
    "ModuleName": "Cities"
  },
  {
    "DiffItemKind": "CommonDiffItem",
    "NodeKind": "Function",
    "NodeAnnotation": "UnwrapOptional",
    "ChildIndex": "1:1:0",
    "LeftUsr": "s:6Cities05ExtraA0P6blibliySQySSGSSSg_SStc1x_tF",
    "LeftComment": "",
    "RightUsr": "",
    "RightComment": "",
    "ModuleName": "Cities"
  },
```
Does changing the end part of the USR from `blibliySQySSGSSSg_SStc1x_tF` to `blibliySSSgAE_SStc1x_tF` fix things?

**Comment by rudkx** [comment]
@nathawes Yes, that turned out to be it. I just needed to update the signature for the protocol from Cities.swift.

**Comment by rudkx** [comment]
@swift-ci Please smoke test

**Comment by rudkx** [comment]
@swift-ci Please test source compatibility

**Comment by rudkx** [comment]
@DougGregor Feel free to merge if you think this looks good.

**Comment by rudkx** [comment]
@DougGregor ping?

**Comment by DougGregor** [comment]
I'm okay with merging this. Sadly it means XFAIL'ing SwiftLint for now due to the IUO there.

**Comment by rudkx** [comment]
@swift-ci Please smoke test

**Comment by rudkx** [comment]
@swift-ci Please test source compatibility

**Comment by ematejska** [comment]
@swift-ci Please test source compatibility

**Comment by ematejska** [comment]
@swift-ci please test

**Comment by ematejska** [comment]
Made a pull request against the source compat repo to xfail swiftlint, https://github.com/apple/swift-source-compat-suite/pull/103

**Pull request merged**